nodejs/eks/nodegroup.ts

```diff
@@ -428,31 +474,81 @@ export class NodeGroupV2 extends pulumi.ComponentResource implements NodeGroupV2
     }
 }
 
-type NodeGroupOptionsCluster = CoreData | Cluster;
+/**
+ * This is a variant of `NodeGroupV2` that is used for the MLC `NodeGroupV2`. We don't just use `NodeGroupV2`,
+ * because we need to accept `ClusterInternal` as the `cluster` arg, so we can correctly pull out `cluster.core`
+ * for use in creating the `NodeGroupV2`.
+ *
+ * @internal
+ */
+export class NodeGroupV2Internal extends pulumi.ComponentResource {
+    public readonly autoScalingGroup!: pulumi.Output<aws.autoscaling.Group>;
+    public readonly extraNodeSecurityGroups!: pulumi.Output<aws.ec2.SecurityGroup[]>;
+    public readonly nodeSecurityGroup!: pulumi.Output<aws.ec2.SecurityGroup>;
+
+    constructor(name: string, args: NodeGroupV2InternalArgs, opts?: pulumi.ComponentResourceOptions) {
+        const type = "eks:index:NodeGroupV2";
+
+        if (opts?.urn) {
+            const props = {
+                autoScalingGroup: undefined,
+                extraNodeSecurityGroups: undefined,
+                nodeSecurityGroup: undefined,
+            };
+            super(type, name, props, opts);
+            return;
+        }
+
+        super(type, name, args, opts);
 
-function isCoreData(arg: NodeGroupOptionsCluster): arg is CoreData {
-    return (arg as CoreData).cluster !== undefined;
+        const group = createNodeGroupV2Internal(name, args, args.cluster.core, this, opts?.provider);
+        this.autoScalingGroup = pulumi.output(group.autoScalingGroup);
+        this.extraNodeSecurityGroups = pulumi.output(group.extraNodeSecurityGroups ?? []);
+        this.nodeSecurityGroup = pulumi.output(group.nodeSecurityGroup);
+        this.registerOutputs({
+            autoScalingGroup: this.autoScalingGroup,
+            extraNodeSecurityGroups: this.extraNodeSecurityGroups,
+            nodeSecurityGroup: this.nodeSecurityGroup,
+        });
+    }
 }
 
+/** @internal */
+export type NodeGroupV2InternalArgs = Omit<NodeGroupV2Options, "cluster"> & { cluster: ClusterInternal };
+
 /**
  * Create a self-managed node group using CloudFormation and an ASG.
  *
  * See for more details:
  * https://docs.aws.amazon.com/eks/latest/userguide/worker.html
  */
 export function createNodeGroup(name: string, args: NodeGroupOptions, parent: pulumi.ComponentResource, provider?: pulumi.ProviderResource): NodeGroupData {
-    const core = isCoreData(args.cluster) ? args.cluster : args.cluster.core;
+    const core = args.cluster instanceof Cluster ? args.cluster.core : args.cluster;
+    return createNodeGroupInternal(name, args, pulumi.output(core), parent, provider);
+}
 
-    if (!args.instanceProfile && !core.nodeGroupOptions.instanceProfile) {
-        throw new Error(`an instanceProfile is required`);
-    }
+function createNodeGroupInternal(
+    name: string,
+    args: Omit<NodeGroupOptions, "cluster">,
+    core: pulumi.Output<pulumi.Unwrap<CoreData>>,
```

Comment:
The main change here is that `core` needs to be typed as `pulumi.Output<pulumi.Unwrap<CoreData>>` rather than just `CoreData` because that's the type that will come back when the `ClusterInternal` is rehydrated.

Same goes for `createNodeGroupV2Internal` and `createManagedNodeGroupInternal`.